File: java/terminal/terminal-server/src/main/java/com/github/fernthedev/terminal/server/backend/BanManager.java
package com.github.fernthedev.terminal.server.backend;

import com.github.fernthedev.server.ClientConnection;
import com.github.fernthedev.server.Server;
import com.github.fernthedev.core.StaticHandler;
import com.google.gson.Gson;
import lombok.NonNull;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class BanManager {

    private static final File bansFile = new File(getCurrentPath(),"banned.json");

    private List<BannedData> banned = new ArrayList<>();

    public boolean isBanned(@NonNull ClientConnection clientConnection) {
        Server.getLogger().info("Checking ban for " + clientConnection);
        if(banned.isEmpty()) load();

        for(BannedData bannedData : banned) {
            Server.getLogger().info("Checking " + clientConnection.getAddress() + " data "+ bannedData);
            if(bannedData.getIp().equals(clientConnection.getAddress())) {
                Server.getLogger().info("Found banned " + bannedData.getIp());
                return true;
            }
        }
        return false;
    }

    public boolean isBanned(String ip) {
        for(BannedData bannedData : banned) {
            if(bannedData.getIp().equals(ip)) {
                return true;
            }
        }
        return false;
    }

    public BanManager() {
        load();
    }

    public void addBan(ClientConnection clientConnection, BannedData bannedData) {
        try(FileWriter writer = new FileWriter(bansFile)) {

            banned.add(bannedData);

            BannedData[] bannedList = banned.toArray(new BannedData[0]);

            writer.write(new Gson().toJson(bannedList));
        } catch (IOException e) {
            e.printStackTrace();
        }
        clientConnection.close();
        load();
    }



    public void load() {
        if(!bansFile.exists()) {
            try {
                bansFile.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        banned.clear();
        BannedData[] bannedDatas = new Gson().fromJson(StaticHandler.getFile(bansFile),BannedData[].class);
        if(bannedDatas != null) {
            banned.addAll(Arrays.asList(bannedDatas));
        }
    }




    private static String getCurrentPath() {
        return Paths.get("").toAbsolutePath().toString();
    }

}

File: java/server/src/main/java/com/github/fernthedev/lightchat/server/security/BanManager.kt
package com.github.fernthedev.lightchat.server.security;

import com.github.fernthedev.config.common.Config;
import com.github.fernthedev.config.common.exceptions.ConfigLoadException;
import com.github.fernthedev.config.gson.GsonConfig;
import com.github.fernthedev.lightchat.server.ClientConnection;
import com.github.fernthedev.lightchat.server.Server;
import com.github.fernthedev.lightchat.server.event.BanEvent;
import lombok.Getter;
import lombok.NonNull;
import org.apache.commons.lang3.SystemUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class BanManager {

    private static final File bansFile = new File(getCurrentPath(),"banned.json");

    private Config<? extends BannedData> bannedDataConfig;
    private final Server server;

    public BanManager(Server server) {
        this.server = server;
        try {
            initConfig();
        } catch (ConfigLoadException e) {
            e.printStackTrace();
        }
    }

    protected void initConfig() throws ConfigLoadException {
        bannedDataConfig = new GsonConfig<>(new BannedData(), bansFile);

        bannedDataConfig.load();
    }

    public boolean isBanned(@NonNull ClientConnection clientConnection) {
        return isBanned(clientConnection.getAddress());
    }

    public boolean isBanned(String ip) {
        for(String bannedData : bannedDataConfig.getConfigData().getIpAddresses()) {
            if(bannedData.equalsIgnoreCase(ip)) {
                return true;
            }
        }
        return false;
    }

    public void ban(String ip) {
        BanEvent banEvent = new BanEvent(true, ip);

        server.getPluginManager().callEvent(banEvent);

        if (banEvent.isCancelled()) return;


        bannedDataConfig.getConfigData().getIpAddresses().add(banEvent.getBannedIP());

        try {
            bannedDataConfig.syncSave();
        } catch (ConfigLoadException e) {
            e.printStackTrace();
        }

        closeAllIPs(ip);
    }

    public void unban(String ip) {
        BanEvent banEvent = new BanEvent(false, ip);

        server.getPluginManager().callEvent(banEvent);

        if (banEvent.isCancelled()) return;


        bannedDataConfig.getConfigData().getIpAddresses().remove(banEvent.getBannedIP());
        try {
            bannedDataConfig.syncSave();
        } catch (ConfigLoadException e) {
            e.printStackTrace();
        }
    }


    protected void closeAllIPs(String ip) {
        server.getPlayerHandler().getChannelMap().forEach((channel, clientConnection) -> {
            if (clientConnection.getAddress().equalsIgnoreCase(ip)) clientConnection.close();
        });
    }



    private static File getCurrentPath() {
        return SystemUtils.getUserDir();
    }

    @Getter
    public static class BannedData {
        private List<String> ipAddresses = new ArrayList<>();
    }
}
